shell/browser/net/resolve_host_function.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2023 Signal Messenger, LLC
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/net/resolve_host_function.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/functional/bind.h"
+#include "base/values.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/storage_partition.h"
+#include "net/base/host_port_pair.h"
+#include "net/base/net_errors.h"
+#include "net/base/network_isolation_key.h"
+#include "net/dns/public/resolve_error_info.h"
+#include "shell/browser/electron_browser_context.h"
+#include "url/origin.h"
+
+using content::BrowserThread;
+
+namespace electron {
+
+ResolveHostFunction::ResolveHostFunction(
+    ElectronBrowserContext* browser_context,
+    std::string host,
+    ResolveHostCallback callback)
+    : browser_context_(browser_context),
+      host_(host),
+      callback_(std::move(callback)) {}
+
+ResolveHostFunction::~ResolveHostFunction() {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  DCHECK(!receiver_.is_bound());
+}
+
+void ResolveHostFunction::Run() {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  DCHECK(!receiver_.is_bound());
+
+  // Start the request.
+  net::HostPortPair host_port_pair(host_, 0);
+  mojo::PendingRemote<network::mojom::ResolveHostClient> resolve_host_client =
+      receiver_.BindNewPipeAndPassRemote();
+  receiver_.set_disconnect_handler(base::BindOnce(
+      &ResolveHostFunction::OnComplete, this, net::ERR_NAME_NOT_RESOLVED,
+      net::ResolveErrorInfo(net::ERR_FAILED),
+      /*resolved_addresses=*/absl::nullopt,
+      /*endpoint_results_with_metadata=*/absl::nullopt));
+  browser_context_->GetDefaultStoragePartition()
+      ->GetNetworkContext()
+      ->ResolveHost(network::mojom::HostResolverHost::NewHostPortPair(
+                        std::move(host_port_pair)),
+                    net::NetworkAnonymizationKey(), nullptr,
```

Comment:
Should we expose any of the ResolveHostParameters? https://source.chromium.org/chromium/chromium/src/+/main:services/network/public/mojom/host_resolver.mojom;l=173-262;drc=8101d6d81854e4f437e0f03f77694a01aaa8df7a

Reply:
Taking a look.

Reply:
I think I can make it work, but what if the configuration for the parameters would change in the future? Can we guarantee API stability?

Reply:
Ah, I guess we'll find out when updating Chromium version :)

Reply:
That's a risk we take with every API we create. Use your judgement to determine which things ought to be exposed and which shouldn't. I also think it would be a reasonable answer to say we can do it later as needed, but make sure to leave room to extend the API.

Reply:
@nornagon done